odoo/addons/base/models/res_currency.py

```diff
@@ -248,26 +281,52 @@ class CurrencyRate(models.Model):
         ('currency_rate_check', 'CHECK (rate>0)', 'The currency rate must be strictly positive.'),
     ]
 
+    def _raise_different_one(self):
+        raise UserError(_("A company can not have a rate different than 1 for its own currency."))
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        self.env['res.company'].browse({
+            vals['company_id'] for vals in vals_list if 'company_id' in vals
+        })._normalize_rate()
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        self.company_id._normalize_rate()
+        return res
+
+    def unlink(self):
+        company = self.company_id
+        res = super().unlink()
+        company._normalize_rate()
+        return res
+
     def _get_latest_rate(self):
-        return self.search([
-            ('currency_id', '=', self.currency_id.id),
-            ('company_id', '=', self.company_id.id or self.env.company.id),
-            ('name', '<=', self.name or fields.Date.today()),
-        ], order="name desc", limit=1)
+        rates = self.currency_id.rate_ids.filtered(
+            lambda x: x.rate and x.company_id == self.company_id and x.name < self.name).sorted('name')
+        return rates and rates[-1] or rates
```

Comment:
```suggestion
        return self.currency_id.rate_ids.filtered(lambda x: (
            x.rate
            and x.company_id == (self.company_id or self.env.company)
            and x.name <= (self.name or field.Date.today())
        )).sorted('name')[-1:]
```

You need to fallback on today otherwise you will have a comparison between date and bool
You need to fallback on the env company because if you didn't specify it, it will be that one by default